lib/appsignal/logger.rb

```diff
@@ -52,27 +88,42 @@ def formatter=(formatter)
     # We support the various methods in the Ruby
     # logger class by supplying this method.
     # @api private
-    def add(severity, message = nil, group = nil)
+    def add(severity, message = nil, group = nil, &block) # rubocop:disable Metrics/CyclomaticComplexity,Metrics/PerceivedComplexity
+      # If we do not need to broadcast to any loggers and the severity is
+      # below the log level, we can return early.
       severity ||= UNKNOWN
-      return true if severity < level
+      return true if severity < level && @loggers.empty?
 
+      # If the logger is silenced, we do not log *or broadcast* messages
+      # below the log level.
+      return true if @silenced && severity < @level
+
+      # Ensure that the block is only run once, even if several loggers
+      # are being broadcasted to.
+      block = BlockOnce.new(&block) unless block.nil?
+
+      # If the group is not set, we use the default group.
       group = @group if group.nil?
-      if message.nil?
-        if block_given?
-          message = yield
-        else
-          message = group
-          group = @group
-        end
-      end
-      return if message.nil?
+
+      did_not_log = true
 
       @loggers.each do |logger|
-        logger.add(severity, message, group)
+        # Loggers should return true if they did *not* log the message.
+        # If any of the broadcasted loggers logs the message, that counts
+        # as having logged the message.
+        did_not_log &&= logger.add(severity, message, group, &block)
       rescue
         nil
       end
 
+      # If the severity is below the log level, we do not log the message.
+      return did_not_log if severity < level
+
+      message = block.call if block && message.nil?
+
+      return if message.nil?
+
+      message = "#{message.class}: #{message.message}" if message.is_a?(Exception)
```

Comment:
Nit: Could be useful to include the first backtrace line if we're logging exceptions.

Reply:
The existing implementation _explicitly_ chose not to do that (like, there's a test for it and everything!) but I agree that makes sense.